Session closed email: rephrase info #7818

sukanta-27 opened the pull request:
Fixes #7818 

Modified `userEmailTemplate-feedbacksSessionClosing` according to the suggestion in the issue.

sukanta-27 commented:
**Query:** The file `userEmailFragment-sessionClosedAdditionalInstructionsFragment.html` also contains the code : 
```html
<p>
  <strong>No action is required if you have already submitted.</strong>
  <br><span style="background-color: #FFFF00">Any request for deadline extensions should be sent to the instructors of the course,</span> not to the TEAMMATES support team.
</p>
```

Should I rephase the info in this file too (and the corresponding testing files), as the other file. 
(Just want to confirm as I do not want to introduce any unwanted changes 😅)

sukanta-27 commented:
@damithc Sir, a reminder. :)

damithc commented:
@TEAMMATES/seniordevs for your advice

whipermr5 commented:
@sukanta-27 Yes; the original issue is about the session closed email so that file should be fixed. The current changes are actually for the session closing email, not the session closed email (but I believe these changes can remain for consistency between the two emails).

sukanta-27 commented:
@damithc  @whipermr5  Thank you very much for the quick response. :) I've started working on it.

sukanta-27 commented on src/main/resources/userEmailTemplate-feedbackSessionClosing.html:
```diff
@@ -66,8 +66,16 @@
+    <strong>problems regarding team/student data</strong> (e.g. wrong team, misspelled name),
+    <br>and <strong>other non-technical queries</strong> about the feedback session :
+    <br>Contact the instructors of the course: ${coOwnersEmails}
+    <li><strong>Technical help regarding TEAMMATES</strong>  (e.g. submission link is not working) :</li>
```
Sure.

wkurniawan07 commented on src/test/java/teammates/test/cases/logic/EmailGeneratorTest.java:
```diff
@@ -473,7 +473,7 @@ private void verifyEmail(EmailWrapper email, String recipient, String subject, S
 
         // check email body for no left placeholders
-        assertFalse(emailContent.contains("${"));
+        //assertFalse(emailContent.contains("${"));
```
> I thought that this should not be the best approach

Why do you think it is not?

> so I commented the line to see what happens, and all the tests passed.

So there is a regression occurring, e.g. I suspect that the placeholder `${coOwnersEmails}` is not getting replaced. This is what you should fix instead of commenting out (essentially removing) tests.

sukanta-27 replied:
>Why do you think it is not?

I meant Commenting the test seemed like a bad approach to handle this error.

>So there is a regression occurring, e.g. I suspect that the placeholder ${coOwnersEmails} is not getting replaced. This is what you should fix instead of commenting out (essentially removing) tests.

Okay, I am working on this.

sukanta-27 replied:
**Update:** This bug got fixed after I changed the `userEmailFragment-sessionClosedAdditionalInstructionsFragment.html` file. All the `EmailGeneratorTest`s are passing now. <br>
Should I push the changes for a review?

sukanta-27 commented on src/main/resources/userEmailTemplate-feedbackSessionClosing.html:
```diff
@@ -66,8 +66,16 @@
+    <br>and <strong>other non-technical queries</strong> about the feedback session :
+    <br>Contact the instructors of the course: ${coOwnersEmails}
+    <li><strong>Technical help regarding TEAMMATES</strong>  (e.g. submission link is not working) :</li>
+    Email TEAMMATES support team at ${supportEmail}.
```
I know this might be a trivial question, but how would I get the email? should I deploy the project in live server and create a session to get the email? :sweat_smile:

wkurniawan07 replied:
That works, but for a much simpler alternative you can simply open this HTML file in a browser and see how it looks.

sukanta-27 replied:
Okay.

sukanta-27 replied:
**sessionClosingEmailForStudent:**<br>
![screenshot 251](https://user-images.githubusercontent.com/10681755/33388800-eaf0c382-d556-11e7-9fd2-e069d23bdf7f.png)

**sessionClosedEmailForStudent:**<br>
![screenshot 252](https://user-images.githubusercontent.com/10681755/33388836-046853b6-d557-11e7-8216-3e56dbe2a3ea.png)

sukanta-27 replied:
**Query:**
Should I remove this line from `userEmailTemplate-feedbackSessionClosing.html` (as the instructor's emails are already provided in the bullet points.): <br>
![screenshot 251 - copy](https://user-images.githubusercontent.com/10681755/33390271-6afea16c-d55b-11e7-890a-fee20eafab69.png)

One other thing that I wanted to ask is that, when I modified `userEmailFragment-sessionClosedAdditionalInstructionsFragment.html` and ran GodMode, four sessionClosed files were modified. But these two lines were not removed. <br>
![screenshot 252](https://user-images.githubusercontent.com/10681755/33389923-5cfd2116-d55a-11e7-976f-5073c6953429.png)
<br>
And when I tried to remove them (I know this is a big No No in GodMode document, but I created a new branch `tinker` and modified the test files), I ran into AssertionError `<expected>`  `<actual>`  error, (The diff was the last line) <br>

>If you need any help or clarification regarding TEAMMATES, please email us at ${support.email}

This line was not in the `<expected>` file but was in the `<actual>` file. (But how is this possible if I changed every test file to match the content in the email, I mean from where those two lines are generating from if it is not in any of the modified files?). Upon running GodMode again, those lines were re-added in the `sessionClosed` files and the tests ran successfully.

wkurniawan07 replied:
> Should I remove this line from `userEmailTemplate-feedbackSessionClosing.html` (as the instructor's emails are already provided in the bullet points.):

I'm for it, @damithc your opinions? Also, there seems to be an excessive amount of line breaks in the text at least for me.

> But how is this possible if I changed every test file to match the content in the email, I mean from where those two lines are generating from if it is not in any of the modified files?

@sukanta-27 take a look at the lines you modified. Do you see "For any non-technical queries ... you can contact your instructor(s) at: ..." and "If you need ..., please email us at ..." as part of the lines you deleted?

damithc replied:
> I'm for it, @damithc your opinions?

Yup, can remove those redundant parts.

> Also, there seems to be an excessive amount of line breaks in the text at least for me.

OK to reduce unnecessary line breaks in the bullet points too.

sukanta-27 replied:
>Yup, can remove those redundant parts.

Okay, I am working on it. @damithc @wkurniawan07 

>take a look at the lines you modified. Do you see "For any non-technical queries ... you can contact your instructor(s) at: ..." and "If you need ..., please email us at ..." as part of the lines you deleted?

I deleted the line "If you need ..., please email us at ..." in the new branch in my machine (Just to experiment :sweat_smile: ) in these files : 

```
sessionClosedEmailForInstructor.html
sessionClosedEmailForStudent.html
sessionClosedEmailTestingSanitizationForInstructor.html
sessionClosedEmailTestingSanitizationForStudent.html
```
and the lines were not in the `userEmailFragment-sessionClosedAdditionalInstructionsFragment.html` file to begin with, still when I ran `EmailGeneratorTest` I got this.
![screenshot 254](https://user-images.githubusercontent.com/10681755/33433301-0e7380ce-d601-11e7-89e2-cf16e04e4131.png)

Anyway, I will figure it out. :smile:

sukanta-27 replied:
@wkurniawan07 I am sorry, I just realized it was a stupid question. It hadn't paid attention to the fact that `userEmailFragment-sessionClosedAdditionalInstructionsFragment.html` is the additional information for the email. The main Template is in the `userEmailTemplate-feedbackSession.html`. The two lines I mentioned were coming from that file. :sweat_smile: 

But now the problem is that when I am rephasing the file `userEmailTemplate-feedbackSession.html` to remove the redundant lines I mentioned all these files are getting modified : 
```
        modified:   src/test/resources/emails/sessionClosedEmailForInstructor.html
        modified:   src/test/resources/emails/sessionClosedEmailForStudent.html
        modified:   src/test/resources/emails/sessionClosedEmailTestingSanitizationForInstructor.html
        modified:   src/test/resources/emails/sessionClosedEmailTestingSanitizationForStudent.html
        modified:   src/test/resources/emails/sessionOpeningEmailForInstructor.html
        modified:   src/test/resources/emails/sessionOpeningEmailForStudent.html
        modified:   src/test/resources/emails/sessionOpeningEmailTestingSanitizationForInstructor.html
        modified:   src/test/resources/emails/sessionOpeningEmailTestingSanitzationForStudent.html
        modified:   src/test/resources/emails/sessionReminderEmailForInstructor.html
        modified:   src/test/resources/emails/sessionReminderEmailForStudent.html

```

For the sessionClosed test files (like `sessionClosedEmailForStudent.html` ) it is providing the **desired** output : 
![screenshot 256](https://user-images.githubusercontent.com/10681755/33437578-6645037e-d60e-11e7-8788-c7560e23b456.png)

But for other emails like `sessionOpeningEmailForInstructor.html` the contact information is not present (As these files are not using any other files for additional information) : 
![screenshot 257](https://user-images.githubusercontent.com/10681755/33437920-6a0a65ca-d60f-11e7-9567-6b9c2cf4b8dc.png)

I am looking for a solution. But as this may take some refactoring, any suggestions will be helpful  :sweat_smile:

sukanta-27 replied:
@wkurniawan07 **Found a solution:**
Changed the current code of [`userEmailTemplate-feedbackSession.html`](https://github.com/TEAMMATES/teammates/blob/master/src/main/resources/userEmailTemplate-feedbackSession.html)
to (below is the `git diff`)
```html
 </p>

 <p>
-  For any non-technical queries (e.g. assigned to the wrong team, misspelled name, request for deadline extension, etc.), you can contact your instructor(s) at: ${coOwnersEmails}.
-</p>
-
-${additionalInstructions}
-<p>
-  If you need any help or clarifications regarding TEAMMATES, please email us at ${supportEmail}.
+  <ul>
+    ${additionalInstructions}
+    <li><strong>Requests for deadline extensions, problems regarding team/student data</strong> (e.g. wrong team, misspelled name),
+    and <strong>other non-technical queries</strong> about the feedback session :
+    <br>Contact the instructors of the course: ${coOwnersEmails}</li>
+    <li><strong>Technical help regarding TEAMMATES</strong> (e.g. submission link is not working) :
+    Email TEAMMATES support team at ${supportEmail}.</li>
+  </ul>
 </p>

```
and `userEmailFragment-sessionClosedAdditionalInstructionsFragment.html` to : 

```html
-<p>
-  <ul>
-    <li><strong>If already submitted:</strong> No action is required.</li>
-    <li><strong>Requests for deadline extensions, problems regarding team/student data</strong> (e.g. wrong team, misspelled name),
-    and <strong>other non-technical queries</strong> about the feedback session :
-    <br>Contact the instructors of the course: ${coOwnersEmails}</li>
-    <li><strong>Technical help regarding TEAMMATES</strong> (e.g. submission link is not working) :
-    Email TEAMMATES support team at ${supportEmail}.</li>
-  </ul>
-</p>
+<li><strong>If already submitted:</strong> No action is required.</li>

```

Now the SessionClosed files look like this : 
![screenshot 256](https://user-images.githubusercontent.com/10681755/33440283-98cf3c2c-d615-11e7-9253-5feef8b0ad75.png)

and the other files: 
```
       sessionOpeningEmailForInstructor.html
       sessionOpeningEmailForStudent.html
       sessionOpeningEmailTestingSanitizationForInstructor.html
       sessionOpeningEmailTestingSanitzationForStudent.html
       sessionReminderEmailForInstructor.html
       sessionReminderEmailForStudent.html
```
has been changed to : 
![screenshot 258](https://user-images.githubusercontent.com/10681755/33440517-2cd1fdb0-d616-11e7-8916-3426ce9d49a1.png)

I think this can work as the above mentioned files has also been rephased to have the bullet points. :smiley:

sukanta-27 commented:
**Update:** Still trying to figure out why the placeholder is not getting replaced.
FWIW this is the `testng-results` of the two failing testcases:

```xml
<?xml version="1.0" encoding="UTF-8"?>
<testng-results skipped="0" failed="2" total="2" passed="0">
  <reporter-output>
  </reporter-output>
  <suite name="Failed suite [Default suite]" duration-ms="4484" started-at="2017-11-29T23:16:42Z" finished-at="2017-11-29T23:16:47Z">
    <groups>
    </groups>
    <test name="Default test(failed)" duration-ms="4484" started-at="2017-11-29T23:16:42Z" finished-at="2017-11-29T23:16:47Z">
      <class name="teammates.test.cases.logic.EmailGeneratorTest">
        <test-method status="PASS" signature="printTestClassHeader()[pri:0, instance:teammates.test.cases.logic.EmailGeneratorTest@1f36e637]" name="printTestClassHeader" is-config="true" duration-ms="16" started-at="2017-11-30T04:46:42Z" finished-at="2017-11-30T04:46:42Z">
          <reporter-output>
          </reporter-output>
        </test-method> <!-- printTestClassHeader -->
        <test-method status="PASS" signature="setupObjectify()[pri:0, instance:teammates.test.cases.logic.EmailGeneratorTest@1f36e637]" name="setupObjectify" is-config="true" duration-ms="547" started-at="2017-11-30T04:46:42Z" finished-at="2017-11-30T04:46:43Z">
          <reporter-output>
          </reporter-output>
        </test-method> <!-- setupObjectify -->
        <test-method status="PASS" signature="setUpGae()[pri:0, instance:teammates.test.cases.logic.EmailGeneratorTest@1f36e637]" name="setUpGae" is-config="true" duration-ms="771" started-at="2017-11-30T04:46:43Z" finished-at="2017-11-30T04:46:44Z">
          <reporter-output>
          </reporter-output>
        </test-method> <!-- setUpGae -->
        <test-method status="PASS" signature="baseClassSetup()[pri:0, instance:teammates.test.cases.logic.EmailGeneratorTest@1f36e637]" name="baseClassSetup" is-config="true" duration-ms="2175" started-at="2017-11-30T04:46:44Z" finished-at="2017-11-30T04:46:46Z">
          <reporter-output>
          </reporter-output>
        </test-method> <!-- baseClassSetup -->
        <test-method status="FAIL" signature="testGenerateFeedbackSessionEmails_testSanitization()[pri:0, instance:teammates.test.cases.logic.EmailGeneratorTest@1f36e637]" name="testGenerateFeedbackSessionEmails_testSanitization" duration-ms="378" started-at="2017-11-30T04:46:46Z" finished-at="2017-11-30T04:46:46Z">
          <exception class="java.lang.AssertionError">
            <full-stacktrace>
              <![CDATA[java.lang.AssertionError: 
	at org.testng.AssertJUnit.fail(AssertJUnit.java:59)
	at org.testng.AssertJUnit.assertTrue(AssertJUnit.java:24)
	at org.testng.AssertJUnit.assertFalse(AssertJUnit.java:41)
	at org.testng.AssertJUnit.assertFalse(AssertJUnit.java:49)
	at teammates.test.cases.BaseTestCase.assertFalse(BaseTestCase.java:134)
	at teammates.test.cases.logic.EmailGeneratorTest.verifyEmail(EmailGeneratorTest.java:476)
	at teammates.test.cases.logic.EmailGeneratorTest.verifyEmailReceivedCorrectly(EmailGeneratorTest.java:492)
	at teammates.test.cases.logic.EmailGeneratorTest.verifyEmailReceivedCorrectly(EmailGeneratorTest.java:482)
	at teammates.test.cases.logic.EmailGeneratorTest.testGenerateFeedbackSessionEmails_testSanitization(EmailGeneratorTest.java:229)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:483)
	at org.testng.internal.MethodInvocationHelper.invokeMethod(MethodInvocationHelper.java:85)
	at org.testng.internal.Invoker.invokeMethod(Invoker.java:659)
	at org.testng.internal.Invoker.invokeTestMethod(Invoker.java:845)
	at org.testng.internal.Invoker.invokeTestMethods(Invoker.java:1153)
	at org.testng.internal.TestMethodWorker.invokeTestMethods(TestMethodWorker.java:125)
	at org.testng.internal.TestMethodWorker.run(TestMethodWorker.java:108)
	at org.testng.TestRunner.privateRun(TestRunner.java:771)
	at org.testng.TestRunner.run(TestRunner.java:621)
	at org.testng.SuiteRunner.runTest(SuiteRunner.java:357)
	at org.testng.SuiteRunner.runSequentially(SuiteRunner.java:352)
	at org.testng.SuiteRunner.privateRun(SuiteRunner.java:310)
	at org.testng.SuiteRunner.run(SuiteRunner.java:259)
	at org.testng.SuiteRunnerWorker.runSuite(SuiteRunnerWorker.java:52)
	at org.testng.SuiteRunnerWorker.run(SuiteRunnerWorker.java:86)
	at org.testng.TestNG.runSuitesSequentially(TestNG.java:1199)
	at org.testng.TestNG.runSuitesLocally(TestNG.java:1124)
	at org.testng.TestNG.run(TestNG.java:1032)
	at org.testng.remote.AbstractRemoteTestNG.run(AbstractRemoteTestNG.java:114)
	at org.testng.remote.RemoteTestNG.initAndRun(RemoteTestNG.java:251)
	at org.testng.remote.RemoteTestNG.main(RemoteTestNG.java:77)
]]>
            </full-stacktrace>
          </exception> <!-- java.lang.AssertionError -->
          <reporter-output>
          </reporter-output>
        </test-method> <!-- testGenerateFeedbackSessionEmails_testSanitization -->
        <test-method status="FAIL" signature="testGenerateFeedbackSessionEmails()[pri:0, instance:teammates.test.cases.logic.EmailGeneratorTest@1f36e637]" name="testGenerateFeedbackSessionEmails" duration-ms="219" started-at="2017-11-30T04:46:46Z" finished-at="2017-11-30T04:46:47Z">
          <exception class="java.lang.AssertionError">
            <full-stacktrace>
              <![CDATA[java.lang.AssertionError: 
	at org.testng.AssertJUnit.fail(AssertJUnit.java:59)
	at org.testng.AssertJUnit.assertTrue(AssertJUnit.java:24)
	at org.testng.AssertJUnit.assertFalse(AssertJUnit.java:41)
	at org.testng.AssertJUnit.assertFalse(AssertJUnit.java:49)
	at teammates.test.cases.BaseTestCase.assertFalse(BaseTestCase.java:134)
	at teammates.test.cases.logic.EmailGeneratorTest.verifyEmail(EmailGeneratorTest.java:476)
	at teammates.test.cases.logic.EmailGeneratorTest.verifyEmailReceivedCorrectly(EmailGeneratorTest.java:492)
	at teammates.test.cases.logic.EmailGeneratorTest.verifyEmailReceivedCorrectly(EmailGeneratorTest.java:482)
	at teammates.test.cases.logic.EmailGeneratorTest.testGenerateFeedbackSessionEmails(EmailGeneratorTest.java:122)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:483)
	at org.testng.internal.MethodInvocationHelper.invokeMethod(MethodInvocationHelper.java:85)
	at org.testng.internal.Invoker.invokeMethod(Invoker.java:659)
	at org.testng.internal.Invoker.invokeTestMethod(Invoker.java:845)
	at org.testng.internal.Invoker.invokeTestMethods(Invoker.java:1153)
	at org.testng.internal.TestMethodWorker.invokeTestMethods(TestMethodWorker.java:125)
	at org.testng.internal.TestMethodWorker.run(TestMethodWorker.java:108)
	at org.testng.TestRunner.privateRun(TestRunner.java:771)
	at org.testng.TestRunner.run(TestRunner.java:621)
	at org.testng.SuiteRunner.runTest(SuiteRunner.java:357)
	at org.testng.SuiteRunner.runSequentially(SuiteRunner.java:352)
	at org.testng.SuiteRunner.privateRun(SuiteRunner.java:310)
	at org.testng.SuiteRunner.run(SuiteRunner.java:259)
	at org.testng.SuiteRunnerWorker.runSuite(SuiteRunnerWorker.java:52)
	at org.testng.SuiteRunnerWorker.run(SuiteRunnerWorker.java:86)
	at org.testng.TestNG.runSuitesSequentially(TestNG.java:1199)
	at org.testng.TestNG.runSuitesLocally(TestNG.java:1124)
	at org.testng.TestNG.run(TestNG.java:1032)
	at org.testng.remote.AbstractRemoteTestNG.run(AbstractRemoteTestNG.java:114)
	at org.testng.remote.RemoteTestNG.initAndRun(RemoteTestNG.java:251)
	at org.testng.remote.RemoteTestNG.main(RemoteTestNG.java:77)
]]>
            </full-stacktrace>
          </exception> <!-- java.lang.AssertionError -->
          <reporter-output>
          </reporter-output>
        </test-method> <!-- testGenerateFeedbackSessionEmails -->
        <test-method status="PASS" signature="tearDownGae()[pri:0, instance:teammates.test.cases.logic.EmailGeneratorTest@1f36e637]" name="tearDownGae" is-config="true" duration-ms="299" started-at="2017-11-30T04:46:47Z" finished-at="2017-11-30T04:46:47Z">
          <reporter-output>
          </reporter-output>
        </test-method> <!-- tearDownGae -->
        <test-method status="PASS" signature="tearDownObjectify()[pri:0, instance:teammates.test.cases.logic.EmailGeneratorTest@1f36e637]" name="tearDownObjectify" is-config="true" duration-ms="0" started-at="2017-11-30T04:46:47Z" finished-at="2017-11-30T04:46:47Z">
          <reporter-output>
          </reporter-output>
        </test-method> <!-- tearDownObjectify -->
        <test-method status="PASS" signature="printTestClassFooter()[pri:0, instance:teammates.test.cases.logic.EmailGeneratorTest@1f36e637]" name="printTestClassFooter" is-config="true" duration-ms="0" started-at="2017-11-30T04:46:47Z" finished-at="2017-11-30T04:46:47Z">
          <reporter-output>
          </reporter-output>
        </test-method> <!-- printTestClassFooter -->
        <test-method status="PASS" signature="remindUserToDisableGodModeIfRequired()[pri:0, instance:teammates.test.cases.logic.EmailGeneratorTest@1f36e637]" name="remindUserToDisableGodModeIfRequired" is-config="true" duration-ms="0" started-at="2017-11-30T04:46:47Z" finished-at="2017-11-30T04:46:47Z">
          <reporter-output>
          </reporter-output>
        </test-method> <!-- remindUserToDisableGodModeIfRequired -->
      </class> <!-- teammates.test.cases.logic.EmailGeneratorTest -->
    </test> <!-- Default test(failed) -->
  </suite> <!-- Failed suite [Default suite] -->
</testng-results>

```

wkurniawan07 commented:
> Still trying to figure out why the placeholder is not getting replaced.

@sukanta-27 It will help if you try to understand how the placeholders are getting replaced in the first place. This is not a black-boxed function provided by some templating engine, if that is what you expected.

sukanta-27 commented:
>It will help if you try to understand how the placeholders are getting replaced in the first place.

@wkurniawan07 Okay, got that.

sukanta-27 commented:
@damithc @wkurniawan07  

**After reducing line breaks and Removing redundant parts:**

![screenshot 255](https://user-images.githubusercontent.com/10681755/33436290-a0c277ec-d60a-11e7-9ce1-339805495421.png)

Is this okay? This is the `userEmailTemplate-feedbackSessionClosing` file.

sukanta-27 commented:
@wkurniawan07 ready for review.

sukanta-27 commented:
@wkurniawan07 A reminder. :sweat_smile:

sukanta-27 commented:
@damithc @wkurniawan07 @whipermr5 As we are rephasing the `feedbackSession` Email Templates in the issue #7818 , I noticed that the following templates still have the old format (Screenshot below) instead of the bullet points for the contact info.: 
```
userEmailTemplate-feedbackSessionPublished.html
userEmailTemplate-feedbackSessionResendAllLinks.html
userEmailTemplate-feedbackSessionUnpublished.html
userEmailTemplate-feedbackSubmissionConfirmation.html
```
![screenshot 264](https://user-images.githubusercontent.com/10681755/33995564-5d4e37b2-e104-11e7-8730-512583affb78.png)

Can I (or rather **should** I ) rephase the info of these files too, so that the feedbackSession Email templates maintain a consistent look?

**If yes :** Should I open another issue for this and send a PR to that issue? (As this issue mainly concerns with the `sessionClosed` email templates?) Or should I push the changes to this PR only?

damithc commented:
@sukanta-27 We can do in in this PR itself (we can change the issue/PR title to match the new scope) if @wkurniawan07 doesn't have any objections.

sukanta-27 commented:
@wkurniawan07 Sir, a reminder.

wkurniawan07 commented:
@sukanta-27 can go ahead and push to this PR.

sukanta-27 commented:
Ready for review.

wkurniawan07 commented on src/main/resources/userEmailFragment-sessionClosedAdditionalInstructionsFragment.html:
```diff
@@ -1,4 +1 @@
-  <strong>No action is required if you have already submitted.</strong>
-  <br><span style="background-color: #FFFF00">Any request for deadline extensions should be sent to the instructors of the course,</span> not to the TEAMMATES support team.
-</p>
+<li><strong>If already submitted:</strong> No action is required.</li>
```
I don't think it's right to have one email fragment with only this line, and especially when this line is also duplicated in some other templates.
Either that, or the file name should be changed to reflect on the current usage.

sukanta-27 replied:
@wkurniawan07 
I kept just this line in this file because otherwise there was a unnecessary line break between "If already submitted " bullet point and the rest of the bullet points. (And I did not include this line in the `feedbackSession` file either as that template is being used to generate other emails as well. Like `feedbackSessionOpening` email. ) 

Yes, this line is also in `userEmailTemplate-feedbackSessionClosing` file.
I had one question when I was rephasing the email Templates and that was, `userEmailTemplate-feedbackSessionClosing` and `userEmailTemplate-feedbackSession` had almost the same content except two lines : 
```
 To submit feedback for the above session, please go to this Web address: ${submitUrl} //This is the feedbackSessionClosing template
 To ${feedbackAction} your feedback for the above session, please go to this Web address: ${submitUrl} //This is the feedbackSubmission template

If already submitted: No action is required  //This is the feedbackSessionClosing template
${additionalInstructions} //This is the feedbackSubmission template
```
**My question is that :** Is there a reason for two separate files?<br>I mean can't we use the `feedbackSession` file to generate the `feedbackSessionClosing` file using this `userEmailFragment-sessionClosedAdditionalInstructionsFragment.html ` file to provide the line `If already submitted: No action is required.` We can also change the name of this file to match the usage as you suggested. (But right now, I don't know how to do this, It may take a little longer than usual to resolve this like a few days, I hope that is not a problem :sweat_smile: )

sukanta-27 replied:
I modified the `EmailGenerator.java` to use `userEmailTemplate-feedbackSession.html` file as the template for `sessionClosingEmailForInstructor` and `sessionClosingEmailForStudent`, so that we don't need a different template for the closing emails. But I haven't commited anything yet and haven't changed the `userEmailFragment-sessionClosedAdditionalInstructionsFragment.html` , as I want your suggestions on this first 

>Is there a reason for two separate files?
I mean can't we use the feedbackSession file to generate the feedbackSessionClosing file using this userEmailFragment-sessionClosedAdditionalInstructionsFragment.html file to provide the line If already submitted: No action is required. We can also change the name of this file to match the usage as you suggested.

wkurniawan07 commented on src/main/resources/userEmailTemplate-feedbackSession.html:
```diff
@@ -63,12 +63,14 @@
+    <br>Contact the instructors of the course: ${coOwnersEmails}</li>
+    <li><strong>Technical help regarding TEAMMATES</strong> (e.g. submission link is not working) :
+    Email TEAMMATES support team at ${supportEmail}.</li>
+  </ul>
```
This entire block is duplicated in 5 other files. See if you can make a template out of it.

sukanta-27 replied:
Okay, I will try to change this and make it a template.

sukanta-27 commented:
@wkurniawan07 Ready for review.

**Outline of Current solution:** 
 **1**. Added the `userEmailFragment-sessionAdditionalContactInformationFragment.html` file which contains the bullet points (Though I don't know if the name of the file is usable or not).
<br> **2**. Replaced the repeated block from the `userEmailTemplates` with `${additionalContactInformation}` placeholder.
<br> **3**. The `userEmailFragment-sessionClosedAdditionalInstructions.html` file is being used by the following files are the line it contains it not being repeated in any other file : 
```
sessionClosedEmailForInstructor.html
sessionClosedEmailForStudent.html
sessionClosedEmailTestingSanitizationForInstructor.html
sessionClosedEmailTestingSanitizationForStudent.html
sessionClosingEmailForInstructor.html
sessionClosingEmailForStudent.html
```
But I didn't rename the file (as suggested https://github.com/TEAMMATES/teammates/pull/8182#discussion_r157347743) because I don't know which name will be accurate. Any suggestions regarding this?

sukanta-27 commented:
@wkurniawan07 A reminder.

sukanta-27 commented:
@wkurniawan07 A reminder. :sweat_smile:

sukanta-27 commented:
@wkurniawan07 
>Why not just hard-code the text in that template to the relevant places?

This solution will work for the template `userEmailTemplate-feedbackSessionClosing.html`, as it is  being used for generating only the `SessionClosing` emails.<br><br>But the `sessionClosed ` emails are getting generated from the `userEmailTemplate-feedbackSession.html` template, which is also being used by two other emails (`sessionOpening` and `sessionReminder` emails).<br> so if I hardcode `<li><strong>If already submitted:</strong> No action is required.</li>` into `userEmailTemplate-feedbackSession.html` file, then the line `If already submitted: No action is required` will also be included in those other emails as well.

**Possible Solutions:**
1. Keep the Additioinal Information fragment as it is.
2. Use `userEmailTemplate-feedbackSessionClosing.html` file to generate both the `sessionClosed` and `sessionClosing` emails, and hardcode the above mention line in this file, In this way we can also remove the file with one line in it. 
3. We can use the line `If already submitted: No action is required` in all the above mentioned emails as well. In this way, we will get rid of the file with only one line, with the least amount of refactoring. 

Which option I should follow? Or any other suggestion beside these?

sukanta-27 commented:
@damithc @wkurniawan07 @whipermr5 Sir, I'll need your suggestion on this (https://github.com/TEAMMATES/teammates/pull/8182#issuecomment-354111479). <br> should I go with option 3?

sukanta-27 commented:
@wkurniawan07 Sir, a reminder.

**Outline of solution :** As suggested in https://github.com/TEAMMATES/teammates/pull/8182#pullrequestreview-85665809, I have removed the `userEmailFragment-sessionClosedAdditionalInstructionsFragment.html` and added the line `<li><strong>If already submitted:</strong> No action is required.</li>` in respective files.

sukanta-27 commented:
Ready for review.

sukanta-27 commented:
@wkurniawan07 Sir, a reminder. This is ready for review. :sweat_smile:

sukanta-27 commented:
@wkurniawan07 A reminder. Ready for review.

wkurniawan07 commented on src/main/java/teammates/logic/api/EmailGenerator.java:
```diff
@@ -681,4 +682,15 @@ private String generateCoOwnersEmailsLine(String courseId) {
+     * @param course Details in {@code course}.
+     * @return A String containing the contact information after replacing the placeholders.
+     */
+    private String fillUpAdditionalContactInformationFragment(CourseAttributes course, String emailBody) {
```
Do you really need the second parameter if it is always `EmailTemplates.FRAGMENT_SESSION_ADDITIONAL_CONTACT_INFORMATION`?

sukanta-27 replied:
Oh yes. I overlooked that. I will fix this. :sweat_smile:

wkurniawan07 commented on src/main/resources/userEmailFragment-sessionAdditionalContactInformationFragment.html:
```diff
@@ -0,0 +1,9 @@
+    <li><strong>Requests for deadline extensions, problems regarding team/student data</strong> (e.g. wrong team, misspelled name),
+    and <strong>other non-technical queries</strong> about the feedback session :
+    <br>Contact the instructors of the course: ${coOwnersEmails}</li>
+    <li><strong>Technical help regarding TEAMMATES</strong> (e.g. submission link is not working) :
```
To make it consistent, there should be no space before `:`. Similarly to the entry two before this.

sukanta-27 replied:
Okay

wkurniawan07 commented on src/main/resources/userEmailTemplate-feedbackSession.html:
```diff
@@ -62,14 +62,9 @@
-<p>
-  For any non-technical queries (e.g. assigned to the wrong team, misspelled name, request for deadline extension, etc.), you can contact your instructor(s) at: ${coOwnersEmails}.
-</p>
+<p>No action is required if you have already submitted.</p>
```
This one should not be here, otherwise there really is no difference between this and the `sessionClosing` template.

Let's examine the three cases where this template is used (based on your claim):
- `sessionClosed` -> nothing can really be done; the submission links will not work any more.
- `sessionOpening` -> it goes without saying that there is something to be done.
- `sessionReminder` -> here "No action is required" is appropriate. Why not make the `sessionReminder` one follow the `sessionClosing` template?

sukanta-27 replied:
Yes, we can do that. I will use `sessionClosing` template to generate `sessionReminder` emails too.

sukanta-27 commented:
@wkurniawan07 All the suggested changes have been made.
This is ready for review.

wkurniawan07 commented on src/main/java/teammates/common/util/Templates.java:
```diff
@@ -42,10 +42,8 @@ public static String populateTemplate(String template, String... keyValuePairs)
-        public static final String USER_FEEDBACK_SESSION_CLOSING =
-                FileHelper.readResourceFile("userEmailTemplate-feedbackSessionClosing.html");
+        public static final String USER_FEEDBACK_SESSION_CLOSING_AND_REMINDER =
+                FileHelper.readResourceFile("userEmailTemplate-feedbackSessionClosingAndReminder.html");
```
`ClosingAndReminder` is too verbose. Just `Reminder` will do since the purpose of the email is to remind the students anyway.

Actually, since this template and the `USER_FEEDBACK_SESSION` template differs only in the "No action required" portion, let's see if you can push this even further: try to make everything come from only one template file and let the "No action required" portion be generated on demand.

sukanta-27 replied:
Okay, I will do that. That would be great, as I was also thinking that there should be a more elegent solution to generate these emails than to keep two templates that differs only in one line.

wkurniawan07 commented on src/main/java/teammates/logic/api/EmailGenerator.java:
```diff
@@ -681,4 +677,15 @@ private String generateCoOwnersEmailsLine(String courseId) {
+     * @param course Details in {@code course}.
+     * @return A String containing the contact information after replacing the placeholders.
+     */
+    private String fillUpAdditionalContactInformationFragment(CourseAttributes course) {
```
`fillUp*` as a method name implies a void method aimed at doing something. Consider `get*` instead.

sukanta-27 replied:
Okay.

sukanta-27 commented:
**Outline of Solution:** At first I thought of adding a boolean variable `isActionRequired` as the second parameter in the `getAdditonal*` method and if the value of the variable in `true` the line `No action required` will be showed. But adding a extra parameter for a line that will be in two types of emails only seemed unnecessary. <br> So instead of that I used `StringBuffer` (in the closing and reminder email generating methods ) to get the `contact information` from the `get*` method and inserted the line `<p>No action is required if you have already submitted.</p>` at the beginning, before converting it back to string and passing it as a parameter.

sukanta-27 commented:
@wkurniawan07 Ready for review.

sukanta-27 commented on src/main/java/teammates/logic/api/EmailGenerator.java:
```diff
@@ -46,6 +46,7 @@
     // feedback action strings
     private static final String FEEDBACK_ACTION_SUBMIT = "submit";
     private static final String FEEDBACK_ACTION_VIEW = "view";
+    private static final String HTML_NO_ACTION_REQUIRED = "<p>No action is required if you have already submitted.</p>\n";
```
Okay.

sukanta-27 commented on src/main/java/teammates/logic/api/EmailGenerator.java:
```diff
@@ -86,12 +87,16 @@
         CourseAttributes course = coursesLogic.getCourse(session.getCourseId());
-        String template = EmailTemplates.USER_FEEDBACK_SESSION.replace("${status}", FEEDBACK_STATUS_SESSION_OPEN);
+        String template = EmailTemplates.USER_FEEDBACK_SESSION
+                .replace("${status}", FEEDBACK_STATUS_SESSION_OPEN);
```
Oops I missed that. It was `USER_FEEDBACK_SESSION_CLOSING_AND_REMINDER` at the last commit, I just erased the `CLOSING_AND_REMINDER` part, thus overlooking this.

sukanta-27 commented on src/main/java/teammates/logic/api/EmailGenerator.java:
```diff
@@ -86,12 +87,16 @@
+        String template = EmailTemplates.USER_FEEDBACK_SESSION
+                .replace("${status}", FEEDBACK_STATUS_SESSION_OPEN);
+        StringBuffer additionalContactInformation = new StringBuffer(getAdditionalContactInformationFragment(course));
+        additionalContactInformation.insert(0, HTML_NO_ACTION_REQUIRED);
```
Right. :man_facepalming:

sukanta-27 commented on src/main/java/teammates/logic/api/EmailGenerator.java:
```diff
@@ -305,15 +305,16 @@ private EmailWrapper generateFeedbackSessionEmailBaseForInstructorReminders(
                                                  ? instructorsLogic.getInstructorsForCourse(session.getCourseId())
                                                  : new ArrayList<InstructorAttributes>();
+        StringBuffer additionalContactInformation = new StringBuffer(getAdditionalContactInformationFragment(course));
+        additionalContactInformation.insert(0, HTML_NO_ACTION_REQUIRED);
```
Okay.

sukanta-27 commented on src/main/resources/userEmailFragment-sessionAdditionalContactInformationFragment.html:
```diff
@@ -0,0 +1,9 @@
+    and <strong>other non-technical queries</strong> about the feedback session:
+    <br>Contact the instructors of the course: ${coOwnersEmails}</li>
+    <li><strong>Technical help regarding TEAMMATES</strong> (e.g. submission link is not working):
+    Email TEAMMATES support team at ${supportEmail}.</li>
```
Okay. I will do this.

sukanta-27 commented on src/main/java/teammates/logic/api/EmailGenerator.java:
```diff
@@ -681,4 +681,15 @@ private String generateCoOwnersEmailsLine(String courseId) {
 
+    /**
+     * Generates additional contact information for User Email Templates.
+     * @param course Details in {@code course}.
```
Okay, I will omit this line.

sukanta-27 commented on src/main/java/teammates/logic/api/EmailGenerator.java:
```diff
@@ -681,4 +681,15 @@ private String generateCoOwnersEmailsLine(String courseId) {
+    /**
+     * Generates additional contact information for User Email Templates.
+     * @param course Details in {@code course}.
+     * @return A String containing the contact information after replacing the placeholders.
```
Okay. I will modify that.

sukanta-27 commented:
@wkurniawan07 All the suggested changes have been made. This is Ready for review.

wkurniawan07 commented on src/main/java/teammates/logic/api/EmailGenerator.java:
```diff
@@ -46,6 +46,8 @@
     private static final String FEEDBACK_ACTION_SUBMIT = "submit";
     private static final String FEEDBACK_ACTION_VIEW = "view";
+    private static final String HTML_NO_ACTION_REQUIRED = "<p>No action is required if you have already submitted.</p>"
+                                                          + System.lineSeparator();
```
Could be better if put this way:
```java
private static final String HTML_NO_ACTION_REQUIRED =
        "<p>No action is required if you have already submitted.</p>" + System.lineSeparator();
```

sukanta-27 replied:
Okay.

wkurniawan07 commented on src/test/resources/emails/sessionClosedEmailForInstructor.html:
```diff
@@ -66,15 +66,18 @@
+      <strong>Requests for deadline extensions, problems regarding team/student data</strong> (e.g. wrong team, misspelled name),
+      and <strong>other non-technical queries</strong> about the feedback session:
+      <br>
+      Contact the instructors of the course: Instructor Not Yet Joined Course 1(instructorNotYetJoinedCourse1@email.tmt), Instructor1 Course1(instructor1@course1.tmt), Instructor3 Course1(instructor3@course1.tmt)
```
Not your fault, but you would notice that between the instructor name and instructor email there is no whitespace (`name(email)` instead of `name (email)`). Let's identify the cause and fix this.

sukanta-27 replied:
Okay. Looking into this.

sukanta-27 replied:
@wkurniawan07 Found the cause. 

```diff
         for (InstructorAttributes coOwner : coOwners) {
             coOwnersEmailsLine
                     .append(coOwner.getName())
-                    .append('(')
+                    .append(" (")
                     .append(coOwner.getEmail())
                     .append("), ");

```
Just to clarify, as this is not related to this issue should I push the changes in this PR only or should I make another issue for this and send a PR to that issue?

sukanta-27 commented:
@wkurniawan07 Ready for review.

sukanta-27 commented:
@wkurniawan07 Sure.

sukanta-27 commented:
@wkurniawan07 
The `sessionClosing` and `sessionReminder` emails look like this : 
(This is the `sessionClosingEmailForInstructor.html` email)<br>
![screenshot 307](https://user-images.githubusercontent.com/10681755/35449602-67af084e-02b5-11e8-8363-0868e23e7ddf.png)

The `sessionClosed` and `sessionOpening` emails look like this.
(this is the `sessionClosedEmailForInstructor.html`)
![screenshot 306](https://user-images.githubusercontent.com/10681755/35449801-f0a863de-02b5-11e8-9644-959bac716a29.png)
<br>`sessionPublishedEmailForInstructor.html` **Screenshot**<br>
![screenshot 308](https://user-images.githubusercontent.com/10681755/35449911-4a14f810-02b6-11e8-9b1e-ffa5850090a0.png)
<br> `sessionSubmissionConfirmationEmailNegativeTimeZone.html` **Screenshot** <br>
![screenshot 309](https://user-images.githubusercontent.com/10681755/35450035-b6128e74-02b6-11e8-8813-88d225bb6868.png)
<br> `sessionUnpublishedEmailForInstructor.html` **Screenshot** <br>
![screenshot 310](https://user-images.githubusercontent.com/10681755/35450115-f54dc482-02b6-11e8-9b2b-0d815a41c012.png)

**Note:** There are two other types of emails that have the old contact info. 
`studentCourseJoinEmail` and `studentCourseRejoinAfterGoogleIdResetEmail`, I didn't rephrased them because this issue was focused on `session` emails.

sukanta-27 commented:
@wkurniawan07 Thank you :smile:
